docs/source/custom-collections.rst

```diff
@@ -0,0 +1,326 @@
+Custom Collections
+==================
+
+For many problems the built-in dask collections (``dask.array``,
+``dask.dataframe``, ``dask.bag``, and ``dask.delayed``) are sufficient. For
+cases where they aren't it's possible to create your own dask collections. Here
+we describe the required methods to fullfill the dask collection interface.
+
+.. note:: This is considered an advanced feature. For most cases the built-in
+          collections are probably sufficient.
+
+Before reading this you should read and underestand:
+
+- :doc:`overview <graphs>`
+- :doc:`graph specification <spec>`
+- :doc:`custom graphs <custom-graphs>`
+
+The Dask Collection Interface
+-----------------------------
+
+To create your own dask collection, you need to fullfill the following
+interface. Note that there is no required base class:
+
+.. method:: __dask_graph__(self)
+
+    The dask graph.
+
+    Returns
+    -------
+    dsk : MutableMapping, None
+        The dask graph. If ``None``, this instance will not be interpreted as a
+        dask collection, and none of the remaining interface methods will be
+        called.
+
+
+.. method:: __dask_keys__(self)
+
+    The output keys for the dask graph.
+
+    Returns
+    -------
+    keys : list
+        A possibly nested list of keys that represent the outputs of the graph.
+        After computation, the results will be returned in the same layout,
+        with the keys replaced with their corresponding outputs.
+
+
+.. staticmethod:: __dask_optimize__(dsk, keys, \*\*kwargs)
+
+    Given a graph and keys, return a new opimized graph.
+
+    This method can be either a ``staticmethod`` or a ``classmethod``, but not
+    an instancemethod.
+
+    If not implemented, defaults to returning the graph unchanged.
+
+    Parameters
+    ----------
+    dsk : MutableMapping
+    keys : list
+    \*\*kwargs
+        Extra keyword arguments forwarded from the call to ``compute`` or
+        ``persist``. Can be used or ignored as needed.
+
+    Returns
+    -------
+    optimized_dsk : MutableMapping
+        The optimized dask graph.
+
+
+.. staticmethod:: __dask_default_get__(dsk, keys, \*\*kwargs)
+
+    The default scheduler ``get`` to use for this object.
+
+    Usually attached to the class as a staticmethod, e.g.
+
+    >>> from dask.threaded import get as threaded_get
+    >>> class MyCollection(object):
+    ...     # Use the threaded scheduler by default
+    ...     __dask_default_get__ = staticmethod(threaded_get)
+
+
+.. method:: __dask_postcompute__(self)
+
+    Finalizer and (optional) extra arguments.
+
+    Used to implement ``dask.compute``.
+
+    Returns
+    -------
+    finalize : callable
+        A function with the signature ``finalize(results, *extra_args)``.
+        Called with the computed keys from ``__dask_keys__`` and any extra
+        arguments as specified in ``extra_args``. Should perform any necessary
+        finalization before returning from ``compute``. For example, the
```

Comment:
Would be nice to say something like "Should return an equivalent in-memory collection."